src/main/js/dashboard-frontend/src/index.tsx

```diff
@@ -4,96 +4,157 @@
+    AppLayout,
+    Box,
+    Flashbar,
+    FlashbarProps,
+    Spinner,
+    Toggle
+} from "@cloudscape-design/components";
 import ServerEndpoint from "./communication/ServerEndpoint";
 import Breadcrumbs from "./navigation/Breadcrumbs";
 import { SERVICE_ROUTE_HREF_PREFIX } from "./util/constNames";
+import {ErrorBoundary} from "react-error-boundary";
+import createPersistedState from 'use-persisted-state';
+import generateUniqueId from "./util/generateUniqueId";
+import {applyMode, Mode} from "@cloudscape-design/global-styles";
+
 
 export var SERVER: ServerEndpoint;
+const apiResource = (websocketError: (m: ReactNode) => void) => {
+    if (!SERVER) {
+        // @ts-ignore
+        SERVER = new ServerEndpoint(window.WEBSOCKET_PORT, window.USERNAME, window.PASSWORD, 5, websocketError);
+    }
+
+    let status = 0;
+    let e: any;
+    const prom = SERVER.initConnections()
+        .then(() => {
+            status = 1;
+        })
+        .catch((ex) => {
+            status = 2;
+            e = ex;
+        });
+
+    return {
+        read() {
+            switch (status) {
+                case 0:
+                    throw prom;
+                case 1:
+                    return null;
+                case 2:
+                    throw e;
+            }
+        }
+    };
+}
 
-interface AppState {
-    flashItems: FlashbarProps.MessageDefinition[];
-    navigationOpen: boolean;
+const Routes = ({apiResource}: {apiResource: any}) => {
+    apiResource.read();
+    return (<Switch>
+        <Redirect
+            exact
+            from={SERVICE_ROUTE_HREF_PREFIX.slice(1, -1)}
+            to="/"
+        />
+        {routes.map((route: any, index: any) => (
+            <Route
+                exact
+                key={index}
+                path={route.routePath}
+                children={<route.main/>}
+            />
+        ))}
+    </Switch>);
 }
 
-class App extends Component<any, AppState> {
-  constructor(props: any) {
-    super(props);
-    // @ts-ignore
-    SERVER = new ServerEndpoint(window.WEBSOCKET_PORT, window.USERNAME, window.PASSWORD, 5, this.websocketError);
+const useNavOpenState = createPersistedState<boolean>("gg.navOpen");
+const useDarkModeState = createPersistedState<boolean>("gg.darkMode");
+export const DefaultContext = createContext<{darkMode?: boolean, addFlashItem?: (i: FlashbarProps.MessageDefinition, removeExisting?: boolean) => void}>({});
+
+
+const AppFunc = () => {
+    const [flashItems, setFlashItems] = useState([] as FlashbarProps.MessageDefinition[]);
+    const [navigationOpen, setNavigationOpen] = useNavOpenState(true);
+    const [darkMode, setDarkMode] = useDarkModeState(window.matchMedia("(prefers-color-scheme: dark)").matches);
+    if (darkMode) {
+        applyMode(Mode.Dark);
+    } else {
+        applyMode(Mode.Light);
+    }
+
+    const addFlashbarItem = (item: FlashbarProps.MessageDefinition, removeExisting: boolean = true) => {
+        item.dismissible = true;
```

Comment:
Could we avoid mutating input data?

```js
   const _item = { dismissible: true, id: generateUniqueId(),  onDismiss: () => {
            setFlashItems((flashItems) => flashItems.filter(i => i.id !== item.id));
        }, ...item}
```

Reply:
Why not mutate? No one is reusing the input data and mutation is orders of magnitude faster than using spread and making new garbage.